File: java/src/main/java/dev/juherr/mobilityid4j/CheckDigitIso.java
package dev.juherr.mobilityid4j;

import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

/**
 * ISO/EMI3 check digit calculator for contract identifiers.
 */
public final class CheckDigitIso {
    private static final Matrix P1 = new Matrix(0, 1, 1, 1);
    private static final Matrix P2 = new Matrix(0, 1, 1, 2);
    private static final Matrix NEG_P2_MINUS_15 = new Matrix(0, 2, 2, 1);
    private static final Matrix[] P1S = powers(P1, 14);
    private static final Matrix[] P2S = powers(P2, 14);
    private static final Map<Character, Matrix> ENCODING = buildEncoding();
    private static final Map<Matrix, Character> DECODING = buildDecoding(ENCODING);

    private CheckDigitIso() {}

    /**
     * Computes the ISO/EMI3 check digit for a contract ID body.
     *
     * @param code identifier payload without computed check digit
     * @return computed check digit
     */
    public static char compute(String code) {
        String normalized = code.toUpperCase(Locale.ROOT);
        if (normalized.length() != P1S.length) {
            throw new IllegalArgumentException("Code must have a length of " + P1S.length);
        }
        if (normalized.length() != P2S.length) {
            throw new IllegalArgumentException("Code must have a length of " + P2S.length);
        }
        for (int i = 0; i < normalized.length(); i++) {
            if (!Ascii.isUpperOrDigit(normalized.charAt(i))) {
                throw new IllegalArgumentException("Code must consist of uppercase ASCII letters and digits");
            }
        }

        Vec t1 = sumEq(normalized, P1S, true);
        Vec t2 = sumEq(normalized, P2S, false).times(NEG_P2_MINUS_15);
        Matrix m15 = new Matrix(t1.v1 & 1, t1.v2 & 1, t2.v1 % 3, t2.v2 % 3);
        Character decoded = DECODING.get(m15);
        if (decoded == null) {
            throw new IllegalStateException("Undecodable matrix: " + m15 + ".");
        }
        return decoded;
    }

    private static Vec sumEq(String code, Matrix[] ps, boolean firstRow) {
        Vec v = new Vec(0, 0);
        for (int i = 0; i < ps.length; i++) {
            char ch = code.charAt(i);
            Matrix encoded = ENCODING.get(ch);
            if (encoded == null) {
                throw new IllegalArgumentException("Invalid character: " + ch + ".");
            }
            Vec row = firstRow ? new Vec(encoded.m11, encoded.m12) : new Vec(encoded.m21, encoded.m22);
            v = v.plus(row.times(ps[i]));
        }
        return v;
    }

    private static Matrix[] powers(Matrix base, int count) {
        Matrix[] result = new Matrix[count];
        Matrix current = base;
        for (int i = 0; i < count; i++) {
            result[i] = current;
            current = current.times(base);
        }
        return result;
    }

    private static Matrix decode(int x) {
        return new Matrix(x & 1, (x >> 1) & 1, (x >> 2) & 3, x >> 4);
    }

    private static Map<Character, Matrix> buildEncoding() {
        Map<Character, Integer> raw = new HashMap<>();
        raw.put('0', 0);
        raw.put('1', 16);
        raw.put('2', 32);
        raw.put('3', 4);
        raw.put('4', 20);
        raw.put('5', 36);
        raw.put('6', 8);
        raw.put('7', 24);
        raw.put('8', 40);
        raw.put('9', 2);
        raw.put('A', 18);
        raw.put('B', 34);
        raw.put('C', 6);
        raw.put('D', 22);
        raw.put('E', 38);
        raw.put('F', 10);
        raw.put('G', 26);
        raw.put('H', 42);
        raw.put('I', 1);
        raw.put('J', 17);
        raw.put('K', 33);
        raw.put('L', 5);
        raw.put('M', 21);
        raw.put('N', 37);
        raw.put('O', 9);
        raw.put('P', 25);
        raw.put('Q', 41);
        raw.put('R', 3);
        raw.put('S', 19);
        raw.put('T', 35);
        raw.put('U', 7);
        raw.put('V', 23);
        raw.put('W', 39);
        raw.put('X', 11);
        raw.put('Y', 27);
        raw.put('Z', 43);

        Map<Character, Matrix> encoded = new HashMap<>();
        raw.forEach((k, v) -> encoded.put(k, decode(v)));
        return Map.copyOf(encoded);
    }

    private static Map<Matrix, Character> buildDecoding(Map<Character, Matrix> encoding) {
        Map<Matrix, Character> reversed = new HashMap<>();
        encoding.forEach((k, v) -> reversed.put(v, k));
        return Map.copyOf(reversed);
    }

    private record Matrix(int m11, int m12, int m21, int m22) {
        Matrix times(Matrix other) {
            return new Matrix(
                    m11 * other.m11 + m12 * other.m21,
                    m11 * other.m12 + m12 * other.m22,
                    m21 * other.m11 + m22 * other.m21,
                    m21 * other.m12 + m22 * other.m22);
        }
    }

    private record Vec(int v1, int v2) {
        Vec plus(Vec other) {
            return new Vec(v1 + other.v1, v2 + other.v2);
        }

        Vec times(Matrix m) {
            return new Vec(v1 * m.m11 + v2 * m.m21, v1 * m.m12 + v2 * m.m22);
        }
    }
}

File: php/src/CheckDigitIso.php
<?php

declare(strict_types=1);

/*
 * This file is part of the Mobility ID library.
 *
 * Copyright (c) 2026 Julien Herr, and respective contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace Juherr\MobilityId;

use Juherr\MobilityId\CheckDigitIso\Matrix;
use Juherr\MobilityId\CheckDigitIso\Vec;

final class CheckDigitIso
{
    private const int LENGTH = 14;

    /**
     * ISO 15118-1 Annex A encoding table: each character maps to a 2x2 matrix over the integers,
     * packed as bits (m11 = bit 0, m12 = bit 1, m21 = bits 2-3, m22 = bits 4+).
     */
    private const array ENCODING = [
        '0' => 0, '1' => 16, '2' => 32,
        '3' => 4, '4' => 20, '5' => 36,
        '6' => 8, '7' => 24, '8' => 40,
        '9' => 2, 'A' => 18, 'B' => 34,
        'C' => 6, 'D' => 22, 'E' => 38,
        'F' => 10, 'G' => 26, 'H' => 42,
        'I' => 1, 'J' => 17, 'K' => 33,
        'L' => 5, 'M' => 21, 'N' => 37,
        'O' => 9, 'P' => 25, 'Q' => 41,
        'R' => 3, 'S' => 19, 'T' => 35,
        'U' => 7, 'V' => 23, 'W' => 39,
        'X' => 11, 'Y' => 27, 'Z' => 43,
    ];

    private function __construct() {} // Prevent instantiation

    /**
     * Calculate ISO check digit.
     *
     * @param string $code The input code string (e.g., "NN123ABCDEFGHI").
     *
     * @return string the calculated check digit character
     *
     * @throws \InvalidArgumentException if the code is malformed
     */
    public static function calculate(string $code): string
    {
        $code = strtoupper($code);
        $codeLength = \strlen($code);

        if ($codeLength !== self::LENGTH) {
            throw new \InvalidArgumentException(
                'Code must have a length of ' . self::LENGTH . ". (Was: {$codeLength})"
            );
        }

        // Check for ASCII uppercase letters and digits
        if (preg_match('/^[A-Z0-9]+$/', $code) !== 1) {
            throw new \InvalidArgumentException(
                "Code must consist of uppercase ASCII letters and digits. (Was: {$code})"
            );
        }

        // The tables are cheap to rebuild (14 matrix products), so nothing is cached: a static
        // cache would only be exercised by the first test of a process and hide table mutations
        // from mutation testing.
        $p1s = self::powers(new Matrix(0, 1, 1, 1));
        $p2s = self::powers(new Matrix(0, 1, 1, 2));
        $negP2Minus15 = new Matrix(0, 2, 2, 1);

        $t1 = self::sumEq($p1s, static fn (Matrix $m): Vec => new Vec($m->m11, $m->m12), $code);
        $t2 = self::sumEq($p2s, static fn (Matrix $m): Vec => new Vec($m->m21, $m->m22), $code)->multiply($negP2Minus15);

        $m15 = new Matrix($t1->v1 & 1, $t1->v2 & 1, $t2->v1 % 3, $t2->v2 % 3);

        // Find the matching char in decoding map
        foreach (self::ENCODING as $char => $value) {
            if ($m15->equals(self::decode($value))) {
                return (string) $char; // Integer-like keys ('0'..'9') are stored as int by PHP
            }
        }

        // Fallback if not found (shouldn't happen with valid input and correct logic)
        throw new \InvalidArgumentException("Undecodable matrix: ({$m15->m11},{$m15->m12},{$m15->m21},{$m15->m22}).");
    }

    /**
     * @return list<Matrix> $base^1 .. $base^LENGTH
     */
    private static function powers(Matrix $base): array
    {
        $powers = [];
        $current = $base;
        for ($i = 0; $i < self::LENGTH; ++$i) {
            $powers[] = $current;
            $current = $current->multiply($base);
        }

        return $powers;
    }

    private static function decode(int $x): Matrix
    {
        return new Matrix(
            $x & 1,
            ($x >> 1) & 1,
            ($x >> 2) & 3,
            $x >> 4
        );
    }

    /**
     * @param list<Matrix>          $ps
     * @param callable(Matrix): Vec $f
     */
    private static function sumEq(array $ps, callable $f, string $code): Vec
    {
        $v = new Vec(0, 0);
        foreach ($ps as $i => $p) {
            $char = $code[$i];
            if (! isset(self::ENCODING[$char])) {
                throw new \InvalidArgumentException("Invalid character: {$char}.");
            }
            $qr = $f(self::decode(self::ENCODING[$char]));
            $v = $v->add($qr->multiply($p));
        }

        return $v;
    }
}
